pkg/broker/ingress/stats_reporter.go

```diff
@@ -0,0 +1,153 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"context"
+	"log"
+	"strconv"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+	"knative.dev/pkg/metrics"
+	"knative.dev/pkg/metrics/metricskey"
+)
+
+// stats_exporter is adapted from knative.dev/eventing/pkg/broker/ingress/stats_reporter.go
+// with the following changes:
+// - Metric descriptions are updated to match GCP broker specifics.
+// - Pod name and container name tags are removed. These can cause high
+// cardinality with little value. Namespace and broker name tags are good enough
+// for identifying the metrics source.
+
+var (
+	// eventCountM is a counter which records the number of events received
+	// by the Broker.
+	eventCountM = stats.Int64(
+		"event_count",
+		"Number of events received by a Broker",
+		stats.UnitDimensionless,
+	)
+
+	// dispatchTimeInMsecM records the time spent dispatching an event to
+	// a Channel, in milliseconds.
+	dispatchTimeInMsecM = stats.Float64(
+		"event_dispatch_latencies",
+		"The time spent dispatching an event to the decouple topic",
+		stats.UnitMilliseconds,
+	)
+
+	// Create the tag keys that will be used to add tags to our measurements.
+	// Tag keys must conform to the restrictions described in
+	// go.opencensus.io/tag/validate.go. Currently those restrictions are:
+	// - length between 1 and 255 inclusive
+	// - characters are printable US-ASCII
+	namespaceKey         = tag.MustNewKey(metricskey.LabelNamespaceName)
+	brokerKey            = tag.MustNewKey(metricskey.LabelBrokerName)
+	eventTypeKey         = tag.MustNewKey(metricskey.LabelEventType)
+	responseCodeKey      = tag.MustNewKey(metricskey.LabelResponseCode)
+	responseCodeClassKey = tag.MustNewKey(metricskey.LabelResponseCodeClass)
+)
+
+type ReportArgs struct {
+	Namespace string
+	Broker    string
+	EventType string
+}
+
+func init() {
+	register()
+}
+
+// StatsReporter defines the interface for sending ingress metrics.
+type StatsReporter interface {
+	ReportEventCount(args *ReportArgs, responseCode int) error
+	ReportEventDispatchTime(args *ReportArgs, responseCode int, d time.Duration) error
+}
+
+var _ StatsReporter = (*reporter)(nil)
+var emptyContext = context.Background()
+
+// Reporter holds cached metric objects to report ingress metrics.
+type reporter struct{}
+
+// NewStatsReporter creates a reporter that collects and reports ingress metrics.
+func NewStatsReporter() StatsReporter {
+	return &reporter{}
+}
+
+func register() {
+	tagKeys := []tag.Key{
+		namespaceKey,
+		brokerKey,
+		eventTypeKey,
+		responseCodeKey,
+		responseCodeClassKey,
+	}
+
+	// Create view to see our measurements.
```

Comment:
It is not necessary to use separate measures for event count and latency, these should be different aggregations of the same mesasure.

Reply:
I am not quite familiar with opencensus. But shouldn't event_count and latency have different type (int vs. float) and units? How can I have a "generic" measure for both?

Reply:
Check out the example on opencensus.io: https://opencensus.io/stats/view/

Reply:
Done.